storage/innobase/btr/btr0cur.cc

```diff
@@ -4169,6 +4174,15 @@ dberr_t btr_cur_pessimistic_update(ulint flags, btr_cur_t *cursor,
   ut_ad(rec_offs_validate(rec, cursor->index, *offsets));
   page_cursor->rec = rec;
 
+  if (was_before_last) {
+    /** We may moved the record from the current page to the next page.
+     * Therefore, we need to disable optimistic access to the next page to
+     * ensure that the record will not be mistakenly skipped during reverse
+     * scanning on cursor.
+     */
+    buf_block_modify_clock_inc(btr_cur_get_block(cursor));
```

Comment:
while PS-9092 blocked the merge which cause the movement of records, this bug shows that records are moved without a merge.

But when a record is moved pessimistically, how can we assumed it is moved to neighbour? I think you should limit this check if movement is neighbour only.

If the record is moved to a non-neightbour, please check this case.
It should be protected by delete marking of record. ie we should read delete marked version of record and the new version which is moved to some far page is not visible/seen. This is OK.

To summarize:
1. record movement due to merges (PS-9092)
2. record movement to neighbour without merge (this PR )
3. record movement to FAR pages (to be verified that delete marked version is read and new version is not SEEN).

Reply:
Thanks, but in our observations, pessimistic updates are limited to the current page and its sibling pages. Whether it is inserted after splitting or inserted directly into its sibling page, the inserted page needs to be disabled for optimistic access, so we do not need stricter judgment here. In addition, the original record has been physically deleted from the original page, and no changes to the delete-mark will be involved.

Reply:
Hmm, thats because the update only changes non-key fields and the new record should always land in neighbour pages.

for your query: update c set val=repeat('b',6000) where id=2;

But if there are queries that update key, records can move with delete marking. 
For example UPDATE c SET id = 20 WHERE id = 5;

Reply:
Whether it is PS-9092 or PS-9297, the necessary condition to trigger the problem is that the same physical record is moved between adjacent pages. But UPDATE c SET id = 20 WHERE id = 5; actually del-marks the record with id=5 and inserts the record with id=20 within a transaction. These two records are not physically the same record. Therefore I think this scenario is not affected by this problem.